src/consensus/dkg.rs

```diff
@@ -368,6 +357,113 @@ struct Participant {
     timer_token: u64,
 }
 
+impl Participant {
+    fn process_message(
+        &mut self,
+        node: &Node,
+        dkg_key: DkgKey,
+        message: DkgMessage,
+    ) -> Result<Vec<Command>> {
+        if self.dkg_key != dkg_key {
+            return Ok(vec![]);
+        }
+
+        let responses = self
+            .key_gen
+            .handle_message(&mut rand::thread_rng(), message)
+            .unwrap_or_default();
+
+        let mut commands = vec![];
+        for response in responses {
+            commands.extend(self.broadcast(node, dkg_key, response)?);
+        }
+
+        Ok(commands)
+    }
+
+    fn broadcast(
+        &mut self,
+        node: &Node,
+        dkg_key: DkgKey,
+        dkg_message: DkgMessage,
+    ) -> Result<Vec<Command>> {
+        let recipients: Vec<_> = self
+            .elders_info
+            .as_ref()
+            .into_iter()
+            .flat_map(EldersInfo::peers)
+            .filter(|peer| *peer.name() != node.name())
+            .map(Peer::addr)
+            .copied()
+            .collect();
+
+        trace!(
+            "broadcasting DKG message {:?} to {:?}",
+            dkg_message,
+            recipients
+        );
+
+        let variant = Variant::DKGMessage {
+            dkg_key,
+            message: bincode::serialize(&dkg_message)?.into(),
+        };
+        let message = Message::single_src(node, DstLocation::Direct, variant, None, None)?;
+
+        let mut commands = vec![];
+        commands.push(Command::send_message_to_targets(
+            &recipients,
+            recipients.len(),
+            message.to_bytes(),
+        ));
+        commands.extend(self.process_message(node, dkg_key, dkg_message)?);
+
+        Ok(commands)
+    }
+
+    fn check(&mut self) -> Option<Command> {
+        if !self.key_gen.is_finalized() {
+            return None;
+        }
+
+        let (participants, outcome) = self.key_gen.generate_keys()?;
+        let elders_info = self.elders_info.take()?;
```

Comment:
is it safe to do so?

Reply:
If `self.elders_info` is `None` that means the DKG has completed. So I think it is. Only issue is we would now still call `generate_keys`. Not sure it's a problem. Maybe just wasted computation? This was solved by the previous approach, but that required `unwrap`.